flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -1552,45 +1583,131 @@ def fir_ArrayCoorOp : fir_Op<"array_coor", [NoSideEffect]> {
     This operation is intended to capture all the runtime values needed to
     compute the address of an array reference in a single high-level op. Given
     the following Fortran input:
-    
+
     ```fortran
       real :: a(n,m)
       ...
       ... a(i,j) ...
     ```
-    
+
     One can use `fir.array_coor` to determine the address of `a(i,j)`.
-    
+
     ```mlir
-      %dim = fir.gendims %one, %n, %one, %one, %m, %one : (...) -> !fir.dims<2>
-      %1 = fir.array_coor %a(%dim), %i, %j : (!fir.ref<!fir.array<?x?xf32>>, ...) -> !fir.ref<f32>
+      %s = fir.shape %n, %m : (index, index) -> !fir.shape<2>
+      %1 = fir.array_coor %a(%s), %i, %j : (!fir.ref<!fir.array<?x?xf32>>, !fir.shape<2>, index, index) -> !fir.ref<f32>
     ```
   }];
 
-  let arguments = (ins AnyReferenceLike:$ref, fir_DimsType:$dims,
-                   Variadic<AnyIntegerType>:$coor);
-		   
+  let arguments = (ins AnyReferenceLike:$memref, Variadic<AnyEmboxArg>:$args);
```

Comment:
Is it possible to define shape and slice as `Optional` arguments, that would also make the custom accessors in `extraClassDeclaration` unnecessary.
Looking at the example it might even be possible to use `assemblyFormat`.
```
class FormatOptionalOperandResultOpBase<string suffix, string fmt>
    : TEST_Op<"format_optional_operand_result_" # suffix # "_op",
              [AttrSizedOperandSegments]> {
  let arguments = (ins Optional<I64>:$optional, Variadic<I64>:$variadic);
  let results = (outs Optional<I64>:$optional_res);
  let assemblyFormat = fmt;
}

def FormatOptionalOperandResultAOp : FormatOptionalOperandResultOpBase<"a", [{
  `(` $optional `:` type($optional) `)` `:` type($optional_res)
  (`[` $variadic^ `]`)? attr-dict
}]>;

def FormatOptionalOperandResultBOp : FormatOptionalOperandResultOpBase<"b", [{
  (`(` $optional^ `:` type($optional) `)`)? `:` type($optional_res)
  (`[` $variadic^ `]`)? attr-dict
}]>;

```

Reply:
Sounds good. For a long time, there wasn't tblgen support for optional. At least, I never found a hint of it after hunting all over for it.